### PR TITLE
CSD02 mandatory tests

### LINKED ISSUE #4: `minimalSecurityAdvisoryDoc.js` and `minimalVexDoc.js` invalid
The files `minimalSecurityAdvisoryDoc.js` and `minimalVexDoc.js` are invalid. Both are missing the mandatory `/vulnerabilites` element. We added test [6.1.27.11](https://github.com/oasis-tcs/csaf/blob/master/csaf_2.0/prose/csaf-v2-editor-draft.md#612711-vulnerabilities) to detect that.

### LINKED ISSUE #7: Adopt 6.1.21
In CDS02, there was a change to test [6.1.21](https://github.com/oasis-tcs/csaf/blob/master/csaf_2.0/prose/csaf-v2-editor-draft.md#6121-missing-item-in-revision-history). The change can be reviewed at https://github.com/oasis-tcs/csaf/pull/443/files. It basically binds the `revision_history` to a start...

### PR DESCRIPTION
**Pull request by domachine**
This PR includes the additional mandatory tests from CSD02.

- Fixes #4
- Fixes #7

### CODE REVIEW

**Comment by tschmidtb51** [comment]
@domachine Do you have a small script that imports the library and validates a file against one test? That would help me to test some edge cases of the implementation.

Just from looking at the code I couldn't spot anything obvious...

**Comment by domachine** [comment]
> @domachine Do you have a small script that imports the library and validates a file against one test? That would help me to test some edge cases of the implementation.
> 
> Just from looking at the code I couldn't spot anything obvious...

Yes, I can throw one together. But if you want, you could add these edge cases to the `documentTests.js` file to include them in the test-suite. Just copy and paste the JSON into the `content` field. Since JSON is also valid JS, no changes are needed.

**Comment by tschmidtb51** [comment]
> Yes, I can throw one together.

I wouldn't mind if you could do that. I would like to use that in the TC's repo to make sure the example files fail the right tests...

**Comment by domachine** [comment]
> > Yes, I can throw one together.
> 
> I wouldn't mind if you could do that. I would like to use that in the TC's repo to make sure the example files fail the right tests...

Here is a git-patch. Save it on your disk and run `git apply <patch-file-path>` from within the `csaf-validator-lib` repo. Then you can run the script using the following command: `./scripts/runTest.js <doc-file> mandatoryTest_6_1_1`. `mandatoryTest_6_1_1` is a sample here. You can insert any test name from `lib/mandatoryTests.js`, `lib/optionalTests.js`, `lib/schemaTests.js` and `lib/schemaTests.js`.

[run_test_script.patch](https://github.com/secvisogram/csaf-validator-lib/files/8396272/run_test_script.txt)

File: lib/mandatoryTests/mandatoryTest_6_1_33.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        productIDsWithVexJustificationCode.add(productID)
+      })
+      flag.group_ids?.forEach((groupID) => {
+        const productIDs = doc.product_tree.product_groups?.find(
+          (group) => group.group_id === groupID
+        )?.product_ids
+        productIDs?.forEach((productID) => {
+          if (
+            !flagReported &&
```
Comment (valentindey) [inline lib/mandatoryTests/mandatoryTest_6_1_33.js:87]: Why do we need this additional check here? Is `flagReported` even needed? If I see it correctly, it will always be the same value as `productIDsWithVexJustificationCode.has(productID)` or am I missing something?

File: lib/mandatoryTests/mandatoryTest_6_1_21.js [hunk 2]
```diff
@@ -18,6 +18,13 @@ export default function mandatoryTest_6_1_21(doc) {
           .map((e) => parseInt(e.number))
       ).keys()
     )
+    if (sortedNumbers.length > 0 && ![0, 1].includes(sortedNumbers[0])) {
+      isValid = false
+      errors.push({
+        instancePath: `/document/tracking/revision_history`,
+        message: `does not start with a version of 0 or 1`,
```
Comment (valentindey) [inline lib/mandatoryTests/mandatoryTest_6_1_21.js:25]: Could we be a little more clear in the message what the error is? Without the context of sorting by date, one might not know what "starting" means.

Side note:
I found it to be somewhat obfuscated that in case of semantic versioning the major version is selected by how `parseInt` works.

File: lib/mandatoryTests/mandatoryTest_6_1_33.js (hunk 1 above)
Comment (domachine) [reply lib/mandatoryTests/mandatoryTest_6_1_33.js:87]: Not exactly. It's a little bit confusing here, since the reporting is based on the flag and not on the product-id. You could refer to the product-id using `product_ids` and `group_ids` which would result in a double report of the flag. I have a test for that. Remove this line and you will see one of them fail.

File: lib/mandatoryTests/mandatoryTest_6_1_30.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        revision.number
+      )
+      if (versioningSchema === null) {
+        versioningSchema = revisionNumberVersioningSchema
+      }
+      if (versioningSchema !== revisionNumberVersioningSchema) {
+        ctx.isValid = false
+        ctx.errors.push({
+          instancePath: `/document/tracking/revisions/${revisionIndex}/number`,
+          message: 'integer and semantic versioning',
```
[the remark this answers is not in the document]
Comment (domachine) [reply lib/mandatoryTests/mandatoryTest_6_1_30.js:60]: I took the wording of the test in the spec here but I think you're right. What do you think @tschmidtb51 ?

File: lib/mandatoryTests/mandatoryTest_6_1_31.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        ) {
+          ctx.isValid = false
+          ctx.errors.push({
+            instancePath: `${path}/${branchIndex}/name`,
+            message: 'version range in product version',
+          })
+        }
+        if (Array.isArray(branch.branches)) {
+          checkBranches({
+            path: `${path}/${branchIndex}`,
```
[the remark this answers is not in the document]
Comment (domachine) [reply lib/mandatoryTests/mandatoryTest_6_1_31.js:76]: Good catch! Thank you very much.

File: lib/mandatoryTests/mandatoryTest_6_1_21.js (hunk 2 above)
Comment (domachine) [reply lib/mandatoryTests/mandatoryTest_6_1_21.js:25]: For the message part: I'm pretty sure there is a better wording 😅 

For the part with `parseInt`: The thing is that using this technique, I do not need to check explicitly if there is semver or numeric versioning in place. But I'm definitely open for improvements, any suggestions?

Comment (valentindey) [reply lib/mandatoryTests/mandatoryTest_6_1_21.js:25]: Maybe sth. like `'revision history does not start with a version of 0 or 1 when sorted by date'` makes it more clear.

I just stumbled over the `parseInt` part initially thinking it doesn't even take care of semver. Checking in detail, I totally get it and see how it is easier. However, a short comment might help future readers of the code to understand more quickly.

File: lib/mandatoryTests/mandatoryTest_6_1_30.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      additionalProperties: true,
+      properties: {
+        tracking: {
+          additionalProperties: true,
+          optionalProperties: {
+            revision_history: {
+              elements: {
+                additionalProperties: true,
+                optionalProperties: {
+                  number: { type: 'string' },
```
Comment (tschmidtb51) [inline lib/mandatoryTests/mandatoryTest_6_1_30.js:18]: Why is `number` an optional property? According to the schema it should be mandatory...

Comment (tschmidtb51) [reply lib/mandatoryTests/mandatoryTest_6_1_30.js:18]: Ah, okay. Now I understand that.

File: lib/mandatoryTests/mandatoryTest_6_1_21.js (hunk 2 above)
Comment (tschmidtb51) [reply lib/mandatoryTests/mandatoryTest_6_1_21.js:25]: Regards message: Maybe, we could write `does not start with major version of 0 or 1`.

File: lib/mandatoryTests/mandatoryTest_6_1_33.js (hunk 1 above)
Comment (tschmidtb51) [reply lib/mandatoryTests/mandatoryTest_6_1_33.js:87]: I had a look into it:
It is actually checking twice: If you just remove line 87 - nothing happens as line 65 would catch that.

Line 65 is definitely needed to prevent double reporting...

**Review by tschmidtb51 — CHANGES_REQUESTED** [review]
Please see my comments.

**Comment by domachine** [comment]
Alright I addressed the issues and rebased the history.

File: lib/mandatoryTests/mandatoryTest_6_1_21.js (hunk 2 above)
Comment (valentindey) [reply lib/mandatoryTests/mandatoryTest_6_1_21.js]: As for the error message, I am fine, so I'm resolving this discussion. @tschmidtb51 please re-open if you miss the `major` keyword.

The comment on `parseInt` makes it more clear :+1:

File: lib/mandatoryTests/mandatoryTest_6_1_33.js (hunk 1 above)
Comment (valentindey) [reply lib/mandatoryTests/mandatoryTest_6_1_33.js]: Now that lines 77-80 and 87-90 are actually the same, it could be simplified further, moving the handling of a `porductID` into the `error` function. But I'm not insisting on that and would be fine with resolving the discussion.

**Review by valentindey — APPROVED** [review]

File: tests/all/documentTests.js [hunk 6]
```diff
@@ -1791,4 +1821,294 @@ export default /** @type {const} */ ([
...
+          product_status: {
+            known_not_affected: ['CSAFPID-9080700', 'CSAFPID-9080701'],
+          },
+        },
+      ],
+    },
+  },
+
+  {
+    title: 'Mandatory Test 6.1.33 skips multiple flags across vulnerabilities',
```
Comment (tschmidtb51) [inline tests/all/documentTests.js:2064]: `Skip` seems to be misleading in that context.
```suggestion
    title: 'Mandatory Test 6.1.33 does not evaluate multiple flags across vulnerabilities',

**Review by tschmidtb51 — COMMENTED** [review]
Just a minor wording change suggested...

**Comment by domachine** [comment]
> Just a minor wording change suggested...

Fixed 🙂

**Review by tschmidtb51 — APPROVED** [review]
LGTM
